cou/steps/hypervisor.py

```diff
@@ -160,6 +160,31 @@ def get_azs(self, target: OpenStackRelease) -> AZs:
 
         return azs
 
+    def _upgrade_plan_sanity_checks(
+        self, target: OpenStackRelease, group: HypervisorGroup
+    ) -> None:
+        """Run sanity checks before generating upgrade plan for hypervisor group.
+
+        :param target: OpenStack codename to upgrade.
+        :type target: OpenStackRelease
+        :param group: HypervisorGroup object
+        :type group: HypervisorGroup
+        """
+        for app in self.apps:
+            if app.name not in group.app_units:
```

Comment:
It would be simpler iterating on the `group.app_units` directly so we don't need the if condition:
```python
for app in goup.app_units.keys():
```

Reply:
Yes, but there could be situation that one group do not contain all apps in hypervisor.

Reply:
I didn't realize that we need the Application class instead of just the name, thanks @rgildein for pointing it out in the sync. But I think if you accept my other comment about directly iterating on Applications for all AZs, then we will not have this problem.

Resolving this thread.